### PR TITLE
JAMES-3586 CL one option for the Cassandra blob store

### PR DESCRIPTION
**Pull request by Arsnael**
Just a beginning, many things left to do...

### CODE REVIEW

File: server/blob/blob-cassandra/pom.xml [hunk 1]
```diff
@@ -65,6 +65,10 @@
             <groupId>${james.groupId}</groupId>
             <artifactId>james-server-util</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>metrics-logger</artifactId>
```
Comment (chibenwa) [inline server/blob/blob-cassandra/pom.xml:70]: Why? MetricFactory API is not enough?

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
+package org.apache.james.blob.cassandra;
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java:1]: Oups

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        int repeatCount = MULTIPLE_CHUNK_SIZE * CHUNK_SIZE;
+        String longString = Strings.repeat("0123456789\n", repeatCount);
+        BlobId blobId = Mono.from(testee().save(testee().getDefaultBucketName(), longString, LOW_COST)).block();
+
+        when(defaultBucketDAO().readPartClOne(blobId, 1)).thenReturn(Mono.empty());
+        when(defaultBucketDAO().readPart(blobId, 1)).thenReturn(Mono.empty());
+
+        assertThatThrownBy(() -> IOUtils.toString(testee().read(testee().getDefaultBucketName(), blobId), StandardCharsets.UTF_8))
+            .isInstanceOf(ObjectStoreException.class)
+            .hasMessageContaining("Missing blob part for blobId");
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java:89]: No then we should fallback to CL QUORUM...

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        int repeatCount = MULTIPLE_CHUNK_SIZE * CHUNK_SIZE;
+        String longString = Strings.repeat("0123456789\n", repeatCount);
+        BlobId blobId = Mono.from(testee().save(testee().getDefaultBucketName(), longString, LOW_COST)).block();
+
+        when(defaultBucketDAO().readPartClOne(blobId, 1)).thenReturn(Mono.empty());
+        when(defaultBucketDAO().readPart(blobId, 1)).thenReturn(Mono.empty());
+
+        assertThatThrownBy(() -> Mono.from(testee().readBytes(testee().getDefaultBucketName(), blobId)).block())
+            .isInstanceOf(ObjectStoreException.class)
+            .hasMessageContaining("Missing blob part for blobId");
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java:104]: No then we should fallback to CL QUORUM...

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java [hunk 5]
```diff
@@ -166,6 +166,23 @@ private boolean isDefaultBucket(BucketName bucketName) {
     }
 
     private Mono<ByteBuffer> readPart(BucketName bucketName, BlobId blobId, Integer partIndex) {
+        if (configuration.isOptimisticConsistencyLevel()) {
+            return readPartClOne(bucketName, blobId, partIndex)
```
Comment (Arsnael) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java:184]: Was not sure if we need to apply it on the read for readPart too? But my guess is yes? Just want to be sure :)

**Review by chibenwa — COMMENTED** [review]
You miss...

 - [ ] Falback mechanisms to QUORUM to avoid all inconsistencies
 - [ ] Documentation of the config option

File: server/blob/blob-cassandra/pom.xml (hunk 1 above)
Comment (Arsnael) [inline server/blob/blob-cassandra/pom.xml:70]: I don't like that... at all... but I had to add something for CassandraBlobStoreFactory that is not in the test package and is being used by other test classes in other modules... Not sure what would be the best way here... or add the metrics as a param in forTesting method?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java [hunk 6]
```diff
@@ -168,7 +182,9 @@ private boolean isDefaultBucket(BucketName bucketName) {
     private Mono<ByteBuffer> readPart(BucketName bucketName, BlobId blobId, Integer partIndex) {
         if (configuration.isOptimisticConsistencyLevel()) {
             return readPartClOne(bucketName, blobId, partIndex)
-                .switchIfEmpty(readPartClDefault(bucketName, blobId, partIndex));
+                .doOnNext(any -> metricClOneHitCount.increment())
```
Comment (Arsnael) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java:185]: Should we use the same metric for hit and miss between readBlob and readPart?

File: server/blob/blob-cassandra/pom.xml (hunk 1 above)
Comment (Arsnael) [reply server/blob/blob-cassandra/pom.xml:70]: https://github.com/apache/james-project/pull/436#discussion_r634214463

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java (hunk 3 above)
Comment (Arsnael) [reply server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java:89]: Read above the when and you will see that I miss with quorum too. It's an override just to keep the test behavior and make it pass. The test you are talking about is below

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java (hunk 4 above)
Comment (Arsnael) [reply server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreClOneTest.java:104]: idem as above

**Comment by Arsnael** [comment]
> Falback mechanisms to QUORUM to avoid all inconsistencies

Are you sure you didn't review a bit fast???

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java (hunk 5 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java:184]: ReadPart is the most critical.

File: server/blob/blob-cassandra/pom.xml (hunk 1 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/pom.xml:70]: How about moving the forTesting stuff in tests, and fix imports of tests that depends on it?

> or add the metrics as a param in forTesting method?

Good solution and likely easier so :+1:

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java (hunk 6 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStoreDAO.java:185]: IMO yes.

**Review by chibenwa — APPROVED** [review]

**Comment by Arsnael** [comment]
Build was green. Did a rebase and squashed commits

**Pull request merged**